add SAML routes with multiple strategies in mind

scottserok opened the pull request:
**PR Summary**
This PR changes the `ActionDispatch::Routing::Mapper#devise_saml_authenticatable` behavior.

**Problem summary**
* Before this change, routes for `devise_saml_authenticatable` would conflict with Devise's own `database_authenticatable` routes when starting up the server:
```
vendor/bundle/ruby/2.2.0/gems/actionpack-4.2.0/lib/action_dispatch/routing/route_set.rb:538:in `add_route': Invalid route name, already in use: 'new_user_session'  (ArgumentError)
You may have defined two routes with the same name using the `:as` option, or you may be overriding a route already defined by a resource with the same naming. For the latter, you can restrict the routes created with `resources` as explained here:
http://guides.rubyonrails.org/routing.html#restricting-the-routes-created
```
* Modifying the default routes directly may impact the behavior of Devise for apps that don't use `database_authenticatable` as per discussion in PR #99. 
* Removing `:saml_authenticatable` from the resource model and setting custom routes to a custom controller per [this wiki](https://github.com/apokalipto/devise_saml_authenticatable/wiki/Supporting-multiple-authentication-strategies) requires the application to implement some of what's already implemented by this library.

**Change Summary**
This change introduces a check when defining routes for `devise_saml_authenticatable` to see if the mapping includes the `database_authenticatable` module. If the module exists, then the named routes are slightly modified to include the `saml` keyword to avoid conflict.  If the module does not exist, then the routes are mapped the same as before this PR.  I did not investigate if other modules would conflict with `devise_saml_authenticatable`, but most issues reported specify `database_authenticatable`, so I believe this should resolve #25 and related issues reported.

I did modify the syntax slightly based on some rubocop feedback such as `:k => v` to `k: v` and `"` to `'`.

Unfortunately I couldn't get specs to work before or after making these changes. I kept getting the following error, and I am too lazy right now to get to the bottom of them considering the change set:
```
An error occurred while loading ./spec/devise_saml_authenticatable/strategy_spec.rb.
Failure/Error: ActiveRecord::Base.logger = Logger.new(nil)

LoadError:
  Error loading the 'postgresql' Active Record adapter. Missing a gem it depends on? pg is not part of the bundle. Add it to your Gemfile.
# /Users/scottserok/.rvm/gems/ruby-2.3.7/gems/activerecord-5.2.2/lib/active_record/connection_adapters/postgresql_adapter.rb:4:in `<top (required)>'
...
# ./spec/rails_helper.rb:10:in `<top (required)>'
...
# ./spec/devise_saml_authenticatable/strategy_spec.rb:1:in `<top (required)>'
...
# ------------------
# --- Caused by: ---
# Gem::LoadError:
#   pg is not part of the bundle. Add it to your Gemfile.
#   /Users/scottserok/.rvm/gems/ruby-2.3.7/gems/activerecord-5.2.2/lib/active_record/connection_adapters/postgresql_adapter.rb:4:in `<top (required)>'
```
To manually test, I'm using this forked version and branch in my app and the routes are generated and the rails server starts without error.
```
$ rake routes | grep saml
                             new_saml_user_session GET      /users/saml/sign_in(.:format)                                                              devise/saml_sessions#new
                                 saml_user_session POST     /users/saml/auth(.:format)                                                                 devise/saml_sessions#create
                         destroy_saml_user_session DELETE   /users/logout(.:format)                                                                    devise/saml_sessions#destroy
                             metadata_user_session GET      /users/saml/metadata(.:format)                                                             devise/saml_sessions#metadata
                     idp_destroy_saml_user_session GET|POST /users/saml/idp_sign_out(.:format)                                                         devise/saml_sessions#idp_sign_out
```
If I removed `:database_authenticatable` from my User model, the routes are:
```
$ rake routes | grep saml
                                  new_user_session GET      /users/saml/sign_in(.:format)                                                              devise/saml_sessions#new
                                      user_session POST     /users/saml/auth(.:format)                                                                 devise/saml_sessions#create
                              destroy_user_session DELETE   /users/logout(.:format)                                                                    devise/saml_sessions#destroy
                             metadata_user_session GET      /users/saml/metadata(.:format)                                                             devise/saml_sessions#metadata
                         idp_sign_out_user_session GET|POST /users/saml/idp_sign_out(.:format)                                                         devise/saml_sessions#idp_sign_out
```

If you want to route to a custom controller, this is possible through the same mechanism Devise provides in `config/routes.rb':
```Ruby
Rails.application.routes.draw do
  devise_for :users, controllers: { saml_sessions: 'users/saml_sessions' }
```
```
$ rake routes | grep saml
                             new_saml_user_session GET      /users/saml/sign_in(.:format)                                                              users/saml_sessions#new
                                 saml_user_session POST     /users/saml/auth(.:format)                                                                 users/saml_sessions#create
                         destroy_saml_user_session DELETE   /users/logout(.:format)                                                                    users/saml_sessions#destroy
                             metadata_user_session GET      /users/saml/metadata(.:format)                                                             users/saml_sessions#metadata
                     idp_destroy_saml_user_session GET|POST /users/saml/idp_sign_out(.:format)                                                         users/saml_sessions#idp_sign_out

scottserok commented:
A few of the builds I checked failed with:
`Bundler could not find compatible versions for gem "bundler":`

adamstegman commented:
I will take a look at the tests! Since there's not a lot of test runs they can fail due to dependency updates ☹️ 

Regarding this change, this seems like a good solution, but I have a couple thoughts:

- I share your discomfort with hard-coding `database_authenticatable`
- Will this break existing applications?

Maybe we can solve both by using a config setting, like `route_helper_prefix` = `saml` to get these exact results, but without the negative effects.

scottserok commented:
Hi @adamstegman 👋 

I'm glad you brought up the configuration setting as an alternate solution! I was debating whether a configuration value or an automatic detection was better and didn't come up with a strong argument either way. I went with the auto detection because I prefer things to just work over having to configure another setting 😄That being said, I do agree that a configuration setting is less risky for existing code bases with various permutations of Devise versions and configurations.  I'm all for not breaking apps so I've updated this PR accordingly. I've also added some comments in the README - feel free to suggest edits.

I've been testing this PR against an old app that's been through several Ruby and Rails upgrades currently using devise-4.4.3 and works as expected outputting named routes per original PR comment when the prefix is `'saml'`.

adamstegman commented:
OK, I updated master with a fix for CI, thanks for your patience! I'll give this code change a shot to see how it feels to use, but it's a great idea!

adamstegman commented on lib/devise_saml_authenticatable.rb:
```diff
@@ -19,6 +19,10 @@ class Engine < Rails::Engine
 # Get saml information from config/saml.yml now
 module Devise
+  # Allow route customization to avoid collision
+  mattr_accessor :route_helper_prefix
```
Could you add a `saml_` prefix to this to avoid collisions with other plugins?

scottserok replied:
I added the `saml_` prefix in the latest commit fd12c0b

adamstegman commented on lib/devise_saml_authenticatable/routes.rb:
```diff
@@ -1,12 +1,23 @@
+        match :destroy, path: mapping.path_names[:sign_out], as: 'destroy', via: mapping.sign_out_via
+        get :metadata, path: 'saml/metadata'
+        match :idp_sign_out, path: 'saml/idp_sign_out', via: %i[get post]
+      end
```
I'll take a look at adding routing specs for this before merging.

scottserok replied:
I imagine the specs could look something like what I have locally but never committed. Here's a gist of the specs hopefully a quick copy & paste will have them working for you: https://gist.github.com/scottserok/6ab54f397319ba9d9e095926dd5352a6 
Thanks for reviewing my PR Adam!

adamstegman replied:
Great start! I had to mess around to get the parameter picked up by the context, since the routes need to be re-drawn, so I ended up using an Engine in the test file.
